src/Microsoft.AspNet.Http/DefaultHttpRequest.cs

```diff
@@ -11,96 +11,41 @@
-        {
-            _context = context;
-            _features = features;
-            ((IFeatureCache)this).SetFeaturesRevision();
-        }
-
-        void IFeatureCache.CheckFeaturesRevision()
-        {
-            if (_cachedFeaturesRevision != _features.Revision)
-            {
-                ResetFeatures();
-            }
-        }
-
-        void IFeatureCache.SetFeaturesRevision()
+        public DefaultHttpRequest(HttpContext context, IFeatureCollection features)
         {
-            _cachedFeaturesRevision = _features.Revision;
+            Initialize(context, features);
         }
 
-        public void UpdateFeatures(IFeatureCollection features)
+        public virtual void Initialize(HttpContext context, IFeatureCollection features)
         {
-            _features = features;
-            ResetFeatures();
+            _context = context;
+            _features = new FeatureReferences<FeatureInterfaces>(features);
         }
 
-        private void ResetFeatures()
+        public virtual void Uninitialize()
         {
-            _request = null;
-            _query = null;
-            _form = null;
-            _cookies = null;
-
-            ((IFeatureCache)this).SetFeaturesRevision();
+            _context = null;
+            _features = default(FeatureReferences<FeatureInterfaces>);
         }
 
-        private IHttpRequestFeature HttpRequestFeature
-        {
-            get { return FeatureHelpers.GetAndCache(this, _features, ref _request); }
-        }
+        public override HttpContext HttpContext => _context;
 
-        private IQueryFeature QueryFeature
-        {
-            get
-            {
-                return FeatureHelpers.GetOrCreateAndCache(
-                    this, 
-                    _features, 
-                    (f) => new QueryFeature(f), 
-                    ref _query);
-            }
-        }
+        private IHttpRequestFeature HttpRequestFeature =>
+            _features.Fetch(ref _features.Cache.Request, f => null);
 
-        private IFormFeature FormFeature
-        {
-            get
-            {
-                return FeatureHelpers.GetOrCreateAndCache(
-                    this,
-                    _features,
-                    this,
-                    (r) => new FormFeature(r),
-                    ref _form);
-            }
-        }
+        private IQueryFeature QueryFeature =>
+            _features.Fetch(ref _features.Cache.Query, f => new QueryFeature(f));
 
-        private IRequestCookiesFeature RequestCookiesFeature
-        {
-            get
-            {
-                return FeatureHelpers.GetOrCreateAndCache(
-                    this, 
-                    _features, 
-                    (f) => new RequestCookiesFeature(f), 
-                    ref _cookies);
-            }
-        }
+        private IFormFeature FormFeature =>
+            _features.Fetch(ref _features.Cache.Form, this, self => new FormFeature(self));
```

Comment:
`self => new FormFeature(self)`

`f` => new RequestCookiesFeature(f)`

Use f in both places?